File: libs/go/client/aws/secrets_manager.go
package aws

import (
	"context"
	"encoding/json"
	"fmt"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
	"go.uber.org/zap"

	"github.com/cyphera/cyphera-api/libs/go/logger" // Assuming logger setup
)

// SecretsManagerClient wraps the AWS Secrets Manager client.
type SecretsManagerClient struct {
	svc *secretsmanager.Client
	cfg aws.Config
}

// NewSecretsManagerClient creates and initializes a new Secrets Manager client.
// It uses the default AWS configuration chain (environment variables, shared config, IAM role).
func NewSecretsManagerClient(ctx context.Context) (*SecretsManagerClient, error) {
	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		return nil, fmt.Errorf("unable to load AWS SDK config: %w", err)
	}

	svc := secretsmanager.NewFromConfig(cfg)

	return &SecretsManagerClient{
		svc: svc,
		cfg: cfg,
	}, nil
}

// GetSecretString fetches a secret string from AWS Secrets Manager using an ARN specified by an environment variable.
// If the ARN environment variable (secretArnEnvVar) is not set or fetching fails,
// it falls back to reading the secret directly from another environment variable (fallbackEnvVar).
// It intelligently handles secrets stored as plain text OR as a JSON object with a single key
// (where the value associated with that key is the desired secret).
// It returns the secret value or an error if both methods fail.
func (c *SecretsManagerClient) GetSecretString(ctx context.Context, secretArnEnvVar string, fallbackEnvVar string) (string, error) {
	secretArn := os.Getenv(secretArnEnvVar)

	// Attempt to fetch from Secrets Manager if ARN is provided
	if secretArn != "" {
		logger.Log.Debug("Attempting to fetch secret from Secrets Manager", zap.String("arnEnvVar", secretArnEnvVar), zap.String("secretArn", secretArn))
		input := &secretsmanager.GetSecretValueInput{
			SecretId: aws.String(secretArn),
		}

		result, err := c.svc.GetSecretValue(ctx, input)
		// Process if fetch was successful AND secret string is not nil/empty
		if err == nil && result.SecretString != nil && *result.SecretString != "" {
			fetchedSecretString := *result.SecretString

			// --- Try parsing as JSON with a single key ---
			var secretJSON map[string]string
			jsonErr := json.Unmarshal([]byte(fetchedSecretString), &secretJSON)

			// If JSON parsing worked AND there's exactly one key
			if jsonErr == nil && len(secretJSON) == 1 {
				// Extract the value from the single key
				for key, value := range secretJSON {
					logger.Log.Info("Successfully fetched secret from Secrets Manager (extracted from single-key JSON)",
						zap.String("secretArn", secretArn),
						zap.String("jsonKey", key),
					)
					return value, nil // Return the extracted value
				}
			}
			// --- End JSON parsing attempt ---

			// If it wasn't single-key JSON or parsing failed, assume it's plain text
			if jsonErr != nil {
				logger.Log.Info("Successfully fetched secret from Secrets Manager (treated as plain text)", zap.String("secretArn", secretArn))
			} else {
				// JSON parsing succeeded but wasn't single-key format
				logger.Log.Warn("Fetched secret from Secrets Manager was JSON but not single-key format, returning raw JSON string",
					zap.String("secretArn", secretArn),
					zap.Int("keyCount", len(secretJSON)),
				)
			}
			return fetchedSecretString, nil // Return the raw fetched string

		}
		// Log the Secrets Manager fetch error but continue to fallback
		logger.Log.Warn("Failed to retrieve secret from Secrets Manager, falling back to env var",
			zap.String("secretArnEnvVar", secretArnEnvVar),
			zap.String("secretArn", secretArn),
			zap.String("fallbackEnvVar", fallbackEnvVar),
			zap.Error(err), // Include the error from Secrets Manager
		)
	} else {
		logger.Log.Debug("Secret ARN environment variable not set, falling back to direct env var",
			zap.String("arnEnvVar", secretArnEnvVar),
			zap.String("fallbackEnvVar", fallbackEnvVar),
		)
	}

	// Fallback to direct environment variable
	secretValue := os.Getenv(fallbackEnvVar)
	if secretValue != "" {
		logger.Log.Info("Using secret value from direct environment variable", zap.String("envVar", fallbackEnvVar))
		return secretValue, nil
	}

	// If both methods fail
	logger.Log.Error("Failed to retrieve secret from both Secrets Manager and direct environment variable",
		zap.String("arnEnvVar", secretArnEnvVar),
		zap.String("fallbackEnvVar", fallbackEnvVar),
	)
	return "", fmt.Errorf("secret not found using ARN env var '%s' or direct env var '%s'", secretArnEnvVar, fallbackEnvVar)
}

// GetSecretJSON fetches a secret from AWS Secrets Manager and unmarshals it into the provided struct.
// It expects the secret stored in Secrets Manager to be a JSON string.
// Falls back to os.Getenv(fallbackEnvVar) if ARN is not set or fetch fails, but assumes the fallback is NOT JSON.
// This is specifically tailored for the RDS secret format.
func (c *SecretsManagerClient) GetSecretJSON(ctx context.Context, secretArnEnvVar string, fallbackEnvVar string, target interface{}) error {
	secretArn := os.Getenv(secretArnEnvVar)
	if secretArn != "" {
		logger.Log.Debug("Attempting to fetch JSON secret from Secrets Manager", zap.String("arnEnvVar", secretArnEnvVar), zap.String("secretArn", secretArn))
		input := &secretsmanager.GetSecretValueInput{
			SecretId: aws.String(secretArn),
		}

		result, err := c.svc.GetSecretValue(ctx, input)
		if err == nil && result.SecretString != nil {
			err = json.Unmarshal([]byte(*result.SecretString), target)
			if err == nil {
				logger.Log.Info("Successfully fetched and parsed JSON secret from Secrets Manager", zap.String("secretArn", secretArn))
				return nil // Success
			}
			logger.Log.Warn("Failed to unmarshal JSON secret from Secrets Manager, falling back",
				zap.String("secretArn", secretArn),
				zap.Error(err),
			)
			// Fall through to fallback if JSON parsing fails
		} else {
			logger.Log.Warn("Failed to retrieve secret from Secrets Manager, falling back",
				zap.String("secretArn", secretArn),
				zap.Error(err), // Include the error from Secrets Manager
			)
			// Fall through to fallback
		}

	} else {
		logger.Log.Debug("Secret ARN environment variable not set, falling back", zap.String("arnEnvVar", secretArnEnvVar))
		// Fall through to fallback
	}

	// Fallback logic (assumes fallbackEnvVar holds a DSN, not JSON)
	// If you intend the fallback to also be JSON, this needs modification.
	fallbackValue := os.Getenv(fallbackEnvVar)
	if fallbackValue != "" {
		// Cannot unmarshal a DSN into the target struct designed for JSON.
		// This indicates a configuration mismatch if fallback is needed for RDS secret.
		logger.Log.Error("Fallback needed for JSON secret, but fallback value is not JSON",
			zap.String("arnEnvVar", secretArnEnvVar),
			zap.String("fallbackEnvVar", fallbackEnvVar),
		)
		return fmt.Errorf("secrets Manager fetch failed for %s, and fallback %s is not JSON parsable", secretArnEnvVar, fallbackEnvVar)
	}

	logger.Log.Error("Failed to retrieve JSON secret from Secrets Manager and no fallback available",
		zap.String("arnEnvVar", secretArnEnvVar),
		zap.String("fallbackEnvVar", fallbackEnvVar),
	)
	return fmt.Errorf("secret not found or parsable using ARN env var '%s' or direct env var '%s'", secretArnEnvVar, fallbackEnvVar)
}

File: apps/delegation-server/src/utils/secrets_manager.ts
import { GetSecretValueCommand, SecretsManagerClient } from "@aws-sdk/client-secrets-manager";
import { logger } from "./utils"; // Assuming logger is in utils

const secretsClient = new SecretsManagerClient({}); // Uses default credential chain (environment, shared config, IAM role)

/**
 * Fetches a secret string from AWS Secrets Manager using an ARN specified by an environment variable.
 * If the ARN environment variable (secretArnEnvVar) is not set or fetching fails,
 * it falls back to reading the secret directly from another environment variable (fallbackEnvVar).
 * It intelligently handles secrets stored as plain text OR as a JSON object with a single key
 * (where the value associated with that key is the desired secret).
 *
 * @param secretArnEnvVar The name of the environment variable holding the AWS Secrets Manager ARN.
 * @param fallbackEnvVar The name of the environment variable holding the direct secret value as a fallback.
 * @returns Promise<string> The secret value.
 * @throws Error if the secret cannot be retrieved from either source.
 */
export async function getSecretValue(secretArnEnvVar: string, fallbackEnvVar: string): Promise<string> {
  logger.info("secretArnEnvVar: ", secretArnEnvVar)
  const secretArn = process.env[secretArnEnvVar];
  logger.info("secretArn: ", secretArn)
  // Attempt to fetch from Secrets Manager if ARN is provided
  if (secretArn) {
    logger.info("secretArn is not empty")
    logger.debug(`Attempting to fetch secret from AWS Secrets Manager for ${secretArnEnvVar}`, { secretArn });
    try {
      logger.info("Attempting to fetch secret from AWS Secrets Manager")
      const command = new GetSecretValueCommand({ SecretId: secretArn });
      logger.info("command: ", command)
      const result = await secretsClient.send(command);
      logger.info("result: ", result)

      if (result.SecretString) {
        logger.info("result.SecretString is not empty")
        const fetchedSecretString = result.SecretString;

        // Try parsing as JSON with a single key
        try {
          logger.info("Attempting to parse as JSON with a single key")
          const secretJSON = JSON.parse(fetchedSecretString);
          if (typeof secretJSON === 'object' && secretJSON !== null) {
            const keys = Object.keys(secretJSON);
            if (keys.length === 1 && typeof secretJSON[keys[0]] === 'string') {
              logger.info(`Successfully fetched and extracted secret from Secrets Manager (single-key JSON) for ${secretArnEnvVar}`, { jsonKey: keys[0] });
              return secretJSON[keys[0]];
            }
          }
          logger.info("Not a JSON or not the expected single-key JSON format, treating as plain text")
        } catch (jsonError) {
          logger.info("Error parsing JSON: ", jsonError)
          // Not a JSON or not the expected single-key JSON format, treat as plain text
          logger.debug(`Secret for ${secretArnEnvVar} is not single-key JSON, treating as plain text.`);
        }
        
        // If it wasn't single-key JSON or parsing failed, assume it's plain text
        logger.info(`Successfully fetched secret from Secrets Manager (plain text) for ${secretArnEnvVar}`);
        return fetchedSecretString;
      }
      logger.info("result.SecretString is empty")
      logger.warn(`SecretString is empty from Secrets Manager for ${secretArnEnvVar}, falling back.`);
    } catch (error) {
      logger.info("Error fetching secret from Secrets Manager: ", error)
      logger.warn(`Failed to retrieve secret from Secrets Manager for ${secretArnEnvVar}, falling back to env var ${fallbackEnvVar}.`, { error, secretArn });
      // Fall through to fallback
    }
  } else {
    logger.info("secretArn is empty")
    logger.debug(`Secret ARN environment variable '${secretArnEnvVar}' not set, falling back to direct env var '${fallbackEnvVar}'.`);
  }

  // Fallback to direct environment variable
  const fallbackSecretValue = process.env[fallbackEnvVar];
  if (fallbackSecretValue) {
    logger.info(`Using secret value from direct environment variable '${fallbackEnvVar}'.`);
    return fallbackSecretValue;
  }

  const errorMessage = `Failed to retrieve secret: '${secretArnEnvVar}' (Secrets Manager) and '${fallbackEnvVar}' (direct env var) are both undefined or failed.`;
  logger.error(errorMessage);
  throw new Error(errorMessage);
} 